### ISSUE TITLE
Question about fixing feature extractor

### ISSUE THREAD

**Issue by ymtoo** [ActiveVisionLab/DFNet#25]
Thanks for the great work!

In the paper:
_In stage two, fixing the histogram-assisted NeRF and the feature extractor G, ..._

Referring to the [code](https://github.com/ActiveVisionLab/DFNet/blob/2c8fa7e324f8d17352ed469a8b793e0167e4c592/script/feature/direct_feature_matching.py#L322), the `model` is responsible for estimating the pose, while the `feat_model` generates the features. Both the pose and features are involved in the loss computation. During backpropagation, the gradients from `loss.backed()` flow through both the `model` and `feat_model`.

Could you point me to the part of the code where the feature extractor in `feat_model` is fixed?

**Comment by chenusc11** [comment]
Hi, although the loss will be backpropagated through the feat_model, the parameters in the feat_model are fixed and will not be updated. 

https://github.com/ActiveVisionLab/DFNet/blob/2c8fa7e324f8d17352ed469a8b793e0167e4c592/script/train.py#L123

**Comment by ymtoo** [comment]
Thanks @chenusc11 

[Doesn't `feat_model.eval()` only change the behavior of some layers such as dropout and batch normalization](https://discuss.pytorch.org/t/does-gradients-change-in-model-eval/127732)? To disable gradient calculations, shouldn't `requires_grad=False` or `torch.no_grad()` be used instead?

Here is a minimum working example to demonstrate this.
```python
>>> x = torch.rand(10,5)
>>> y = torch.rand(10,1)
>>> model = nn.Linear(5,1)
>>> model.eval()
Linear(in_features=5, out_features=1, bias=True)
>>> loss_function = lambda x, y : (model(x) - y).abs().sum()
>>> optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
>>> model.weight
Parameter containing:
tensor([[-0.2051, -0.0533,  0.1585, -0.2878, -0.1016]], requires_grad=True)
>>> loss = loss_function(x, y)
>>> loss.backward()
>>> optimizer.step()
>>> model.weight
Parameter containing:
tensor([[-0.1579, -0.0045,  0.2045, -0.2340, -0.0499]], requires_grad=True) # the weights changed
```

**Comment by ymtoo** [comment]
It seems like the optimizer only [updates `model.parameters()` but not `feat_model.parameters()`](https://github.com/ActiveVisionLab/DFNet/blob/2c8fa7e324f8d17352ed469a8b793e0167e4c592/script/train.py#L127).

**Comment by chenusc11** [comment]
Hi, thank you for this question. It's been a while since I played with this code. Please give me some time to verify this. Meanwhile, when you run the code, does the feat_model's weights have requires_grad=True?

**Comment by ymtoo** [comment]
> Hi, thank you for this question. It's been a while since I played with this code. Please give me some time to verify this. Meanwhile, when you run the code, does the feat_model's weights have requires_grad=True?

For 
````bash
python train.py --config config_dfnetdm.txt
````
, `print(feat_model.encoder[0].weight.requires_grad)` returns `True`.

**Comment by chenusc11** [comment]
Hmm... Then I think you are right. The requires_grad here should be set to False. Thank you for your observation! 

The lr was set to very small values in my code when doing DFM. However, feat_model's weights should not be updated, and I will update this fix. Many thanks

**Comment by ymtoo** [comment]
Even though `feat_model.encoder[0].weight.requires_grad=True`, the feat_model's weights should not be updated because [the optimizer only contains the model's parameters](https://github.com/ActiveVisionLab/DFNet/blob/2c8fa7e324f8d17352ed469a8b793e0167e4c592/script/train.py#L127) and no weight update is being done on feat_model.

**Comment by chenusc11** [comment]
Then I think it’s fine :)

**Comment by ymtoo** [comment]
Thank you!
